src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -873,6 +873,127 @@
          :body (get-code-content environment code-id file-path)}))))
 
+(defn is-bolt-project?
+  [dir]
+  (or (fs/exists? (str dir "/bolt-project.yaml"))
+      (fs/exists? (str dir "/bolt.yaml"))
+      (fs/exists? (str dir "/Boltdir"))))
+
+(defn find-bolt-project
+  [project-dir project-name]
+  (->> project-dir
+       io/file
+       .list
+       seq
+       (filter #(is-bolt-project? (str project-dir "/" %)))
+       (some #{project-name})))
+
+(defn get-project-root
+  [project-dir project-name]
+  (let [boltdir-path (str project-dir "/" project-name "/Boltdir")]
+    (if (fs/exists? boltdir-path)
+      boltdir-path
+      (str project-dir "/" project-name))))
+
+(defn- dirs-in-project-modulepath
+  "List all directories in a bolt project's modulepath"
+  [project-root]
+  (->> ["modules" "site-modules" "site"]
+       (map #(str project-root "/" %))
+       (filter fs/exists?)
+       (mapcat fs/list-dir)
+       (filter fs/directory?)))
+
+(defn find-project-module
+  "Given the path of a project and the name of a module, search the project's
+  module path and return the path to that module, or nil."
+  [project-root module]
+  (->> project-root
+       dirs-in-project-modulepath
+       (filter #(= module (fs/base-name %)))
+       first))
+
+(defn mount->path-component
+  "Map the fileserving \"mount\" to the subdirectory of modules it will serve
+  files out of."
+  [mount]
+  (case mount
+    "modules" "files"
+    ;; default to itself
+    mount))
+
+(defn find-project-file
+  [bolt-project-dir project mount module path]
+  (if (is-bolt-project? (str bolt-project-dir "/" project))
+    (let [project-root (get-project-root bolt-project-dir project)
+          module-root (find-project-module project-root module)
+          file-path (str module-root "/" (mount->path-component mount) "/" path)]
+      (if (fs/exists? file-path)
+        (fs/file file-path)))))
+
+(def project-routes
+  "Bidi routing table for project file_content endpoint. This is done separately
+  because we have to do additional routing after branching on the query
+  parameter, which is not natively supported in bidi."
+
+  ["" {[[#"tasks|modules" :mount-point] "/" :module "/" [#".*" :file-path]] :basic
+       [[#"plugins|pluginfacts" :mount-point] [#".*" :file-path]] :pluginsync}])
+
+(defn make-file-content-response
+  "Given a File, generate an appropriate ring response map. Returns a 404
+  response if passed `nil`."
+  [file requested-file-path]
+  (if file
+    (-> file
+        str
+        rr/file-response
+        (rr/content-type "application/octet-stream"))
+    {:status 404
+     :headers {"Content-Type" "text/plain"}
+     :body (i18n/tru "Could not find file_content for path: {0}" requested-file-path)}))
+
+(defn handle-project-file-content
+  "Handle a file_content request for a bolt project."
+  [bolt-project-dir request]
+  (let [project (get-in request [:params "project"])
+        version (get-in request [:params "version"])
```

Comment:
Where does `version` come from?

Reply:
leftover line from an alternate implementation where i treated the different components of the project name independently (which is still something we might choose to do later, not hard to change)